tags/src/lib.rs

```diff
@@ -239,25 +264,59 @@ impl TagsConfiguration {
     }
 }
 
-impl TagsContext {
+impl<'a> TagsContext {
     pub fn new() -> Self {
         TagsContext {
             parser: Parser::new(),
             cursor: QueryCursor::new(),
         }
     }
 
-    pub fn generate_tags<'a>(
+    pub fn parser(&mut self) -> &mut Parser {
+        &mut self.parser
+    }
+
+    pub fn generate_tags(
+        &'a mut self,
+        config: &'a TagsConfiguration,
+        source: &'a [u8],
+        cancellation_flag: Option<Arc<AtomicUsize>>,
+    ) -> Result<(impl Iterator<Item = Result<Tag, Error>> + 'a, bool), Error> {
```

Comment:
I have the same question here as I had about the `Highlighter`. Is it necessary to change this API? The existing API was already safe (because it maintains the contract required by the unsafe `set_cancellation_flag` API).

Reply:
I've decided to write more tests and will come back soon with a bunch on tests that would help to explain things better.
Also I found and tested a better approach how to implement changes suggested in this PR, also planning to try to finish it on the next week.